### PR TITLE
Enable OPI staging

### PR DESCRIPTION
**Pull request by gdankov**
The previous [PR](https://github.com/cloudfoundry/cloud_controller_ng/pull/1205) introduced the ability to run apps in Kubernetes. This one expands on that and allows for staging to run as a Kubernetes job.

  Additionally, we made some other small changes to the apps client in order to get CATs passing.

* [x] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `master` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [x] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)
We deploy Eirini using Kubernetes (instead of Diego) and run CATs apps suite on a regular basis in our [pipeline](https://ci.flintstone.cf.cloud.ibm.com/teams/eirini/pipelines/acceptance/jobs/run-core-cats/builds/12). 

cc @suhlig @julz @JulzDiverse @mnitchev

### CODE REVIEW

**Comment by mnitchev** [comment]
We looked into the [failing Codeclimate check](https://codeclimate.com/github/cloudfoundry/cloud_controller_ng/pull/1288) and it seems to run a different set of Rubocop rules. Locally running `bundle exec rake rubocop` yields no findings, yet Codeclimate complains.

Most of the issues made sense to us, so we fixed them in the last commit. We are trying to understand the 250 lines limit that Codeclimate complains about though. It doesn’t make sense for us to split up the test to fit in this limitaion. Also, we see that there are plenty of files in CC that have many more lines. Maybe you want to enforce this rule for new code? Is there a way to disable this check for the test file?

@JulzDiverse & @mnitchev

File: lib/cloud_controller/opi/apps_client.rb [hunk 1]
```diff
@@ -92,29 +102,64 @@ def routes(process)
...
+    def ports(process)
+      ::VCAP::CloudController::Diego::Protocol::OpenProcessPorts.new(process).to_a
+    end
+
+    def generate_volume_mounts(process)
+      app_volume_mounts   = VCAP::CloudController::Diego::Protocol::AppVolumeMounts.new(process.app).as_json
+      proto_volume_mounts = []
+
+      app_volume_mounts.each do |volume_mount|
+        if volume_mount['device']['mount_config'].present? && volume_mount['device']['mount_config']['name'].present?
```
Comment (ericpromislow) [inline lib/cloud_controller/opi/apps_client.rb]: From what I can tell, this test always fails as `volume_mount['device']['mount_config']` is a string (see https://github.com/cloudfoundry/cloud_controller_ng/blob/3919b14ea002ef62c7869a7900d6834cc222e118/lib/diego/bbs/models/volume_mount.pb.rb#L31 ) and so your doing `STRING['name']`. If `STRING` contains the substring anywhere, this returns `"name"` and test test succeeds (and the `volume_id` field will be set to `"name"`. If not, the test fails.

File: lib/cloud_controller/opi/stager_client.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+module OPI
+  class StagerClient
+    def initialize(opi_url, config)
+      @client = HTTPClient.new(base_url: URI(opi_url))
+      @config = config
+    end
+
+    def stage(staging_guid, staging_details)
+      staging_request = to_request(staging_guid, staging_details)
```
Comment (ericpromislow) [inline lib/cloud_controller/opi/stager_client.rb]: You should probably carry the same `logger.info` messages here that are in `BbsStagerClient#stage` https://github.com/cloudfoundry/cloud_controller_ng/blob/3919b14ea002ef62c7869a7900d6834cc222e118/lib/cloud_controller/diego/bbs_stager_client.rb#L15 — when staging fails, these two messages are useful.

File: spec/integration/opi_apps_client_spec.rb [hunk 3]
```diff
@@ -43,12 +43,20 @@ def up?(url)
...
       let(:droplet) {
-        VCAP::CloudController::DropletModel.new(
+        VCAP::CloudController::DropletModel.make(
           docker_receipt_image: 'http://example.org/image1234',
           droplet_hash: 'd_haash',
           guid: 'some-droplet-guid'
         )
       }
+      let(:lifecycle_type) { nil }
+      let(:app_model) {
```
Comment (ericpromislow) [inline spec/integration/opi_apps_client_spec.rb]: `app_model` is never referenced, and therefore not instantiated. How were you expecting to use it?

**Comment by ericpromislow** [comment]
We currently don't make codeclimate results make-or-break; I have a stack-overflow question out on how to protect pieces of code from codeclimate complaints.  However I did put three questions/comments inline in the `Files changed` tab.  Please review.

**Comment by ericpromislow** [comment]
Also a deploy with opi disabled (default) fails for `api` and `cc-worker` with the following error in pre-start: ```
[2019-02-15 01:38:02+0000] rake aborted!
[2019-02-15 01:38:02+0000] Membrane::SchemaValidationError: { opi => { opi_staging => Missing key, cc_uploader_url => Missing key } }
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/gem_home/ruby/2.4.0/gems/membrane-1.1.0/lib/membrane/schemas/record.rb:81:in `fail!'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/gem_home/ruby/2.4.0/gems/membrane-1.1.0/lib/membrane/schemas/record.rb:60:in `validate'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/gem_home/ruby/2.4.0/gems/membrane-1.1.0/lib/membrane/schemas/record.rb:23:in `validate'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/cloud_controller_ng/lib/vcap/config.rb:17:in `from_file'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/cloud_controller_ng/lib/cloud_controller/config.rb:24:in `load_from_file'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/cloud_controller_ng/lib/tasks/rake_config.rb:12:in `config'
[2019-02-15 01:38:02+0000] /var/vcap/data/packages/cloud_controller_ng/0c6b90c440970500b412a1158ad17594697d5c6d/cloud_controller_ng/lib/tasks/db.rake:109:in `block (2 levels) in <top (required)>'
[2019-02-15 01:38:02+0000] /var/vcap/packages/cloud_controller_ng/gem_home/ruby/2.4.0/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
[2019-02-15 01:38:02+0000] /var/vcap/packages/ruby-2.4-r5/bin/bundle:30:in `block in <main>'
[2019-02-15 01:38:02+0000] /var/vcap/packages/ruby-2.4-r5/bin/bundle:22:in `<main>'
[2019-02-15 01:38:02+0000] Tasks: TOP => db:seed
```

File: lib/cloud_controller/opi/apps_client.rb (hunk 1 above)
Comment (paulcwarren) [reply lib/cloud_controller/opi/apps_client.rb]: Hi @ericpromislow .  We spent some time looking at this today.  

As we understand it the volume mounts being parsed here are the volume mounts that were received by cloud controller in response to a `bind-service` operation on a volume service broker.  If we look at the open service broker api, mount_config is [specified as an object](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#device-object), not a string.   These volume mounts are stored encrypted in the capi database before being used to desire an LRP.   Later on when the app is desired this info is retrieved and decrypted.  

In the case that capi is talking to diego the volume mounts are transformed into the bbs model you reference and as part of that the `mount_config` is [serialized to a string](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/diego/app_recipe_builder.rb#L161).  

In the case that capi is talking to OPI the volume mounts are parsed down to a much simpler model of just the volume name and mount path (i.e. destination of the volume mount inside the eventual container).  For context, the eirini team didn't want the full volume mount specification as used by diego/bbs today.

Either way, the info being deserialized out of the database is a hash and therefore we believe this code is ok.  We have restested it again today and confirmed that we see correct values everywhere.    

Please let us know if you see it differently we are not experts in CAPI, or OPI/Eirini.

**Comment by mnitchev** [comment]
Hi @ericpromislow 

Thanks for the feedback! We made the requested changes (apart from the volume_mounts one, as we delegated to the persistence team). We also fixed a few things that were failing on our side. 
Please let us know if there is anything else we need to change.

BR,
@mnitchev @JulzDiverse

**Comment by julz** [comment]
bump 🙏

File: Gemfile.lock [hunk 4]
```diff
@@ -196,9 +196,8 @@ GEM
       mime-types (~> 3.0)
       representable (~> 3.0)
       retriable (>= 2.0, < 4.0)
-    google-protobuf (3.6.1)
-    google-protobuf (3.6.1-universal-darwin)
```
Comment (tcdowney) [inline Gemfile.lock]: Why was this changed?

Comment (selzoc) [reply Gemfile.lock]: We manually corrected this after the merge in https://github.com/cloudfoundry/cloud_controller_ng/commit/924c904d9a7ca948407a102685ba215f3a5eff20.
